### ISSUE TITLE
[bug] The standard Windows installation has bad UX for installing WebView2 as a dependency

### ISSUE THREAD

**Issue by TapGhoul** [tauri-apps/tauri#4389]
### Describe the bug

See the video below for why I'm ringing the bell about this:

https://user-images.githubusercontent.com/3120639/174429998-b17eb281-c50a-49a9-9118-4193be949678.mp4

When installing a standard Tauri application when the user does not have webview2 installed, the installer behaves like a malware-ridden installer.

The wix installer needs to be made to make it synchronous.

This was previously talked about in https://github.com/tauri-apps/tauri/issues/2452#issuecomment-918291043

It recently got a reply in the discord:

> 1) Personally i prefer a non-quiet installation
> 2) UAC is not an issue anymore because ms added support for installing it with just user permissions
> 3) i still agree that running the wv2 installer at the end/after the app installation is too late
> ...
> it's indeed quite some time ago already so i'm not 100% sure, but i think we don't even need to use powershell but could use wix' ExePackage instead.
> Unless setting tls1.2 in powershell works even if its not enabled system wide, because that would fix the installer on win7. and changing this setting for exepackage is not possible afaik.

https://discord.com/channels/616186924390023171/731495028677148753/985618974882541578

but needs some action taken on it.

If there's discussion on how it should be implemented, I'm happy to send a PR. My personal opinions on how it should work don't match that of the Tauri WG, so I'm not going to do it without some information, but I'm happy to do cross-testing for 7, 10 and 11 to make sure the results work and is a pleasant installation experience, as long as the parameters are laid out.

### Reproduction

1. Build a standard tauri application
2. Remove Webview2 to cause the installer to install it for you
3. Install your build, watch it behave in a creepy fashion

### Platform and versions

```shell
Windows pre-11, no WebView2 installed.
```

**Comment by TapGhoul** [comment]
~~As a note, even down to the UAC prompt (if it remains) - that needs to be tidied up too. Show something like "${AppName} installer" or something instead of the full path to the installer. Again, happy to do all of this, just need info on what is actually required.~~

Edit: I'm dumb, this is a codesigning thing.

**Comment by TapGhoul** [comment]
Personally, I believe the webview2 installer should be silent, not explicit (along with being made synchronous/given its own page) - my reasoning is that the MS office installer makes it silent, as does teams. I see no reason to bother the user when it's a shared component that's becoming standard, as more popups will scare the average user, and Edge already has enough stigma in the public consciousness that stating this loud and proud will make users want to run away.

Especially when they already have Edge installed, and don't know what a webview is.

Most tauri apps will not be used by devs, they'll be used by regular users. The installer should reflect that.

**Comment by FabianLars** [comment]
I'm gotta say, i'm a little confused, i somewhat waited for the PR after our small convo 😅 
I generally agreed with you and even mentioned a wix native alternative i'd like to see being explored to potentially remove the powershell portion.
And yes i said i **personally** don't like silent installers, but that's _my_ opinion not from the team and either way doesn't mean that it shouldn't be explored. But guess what, i changed my mind* 🤷 Do you think there is a way to integrate the download/install progress into the wix progress bar? Probably not, but that does sound fancy.

\* but only if it actually works fine: We had a few problems with silent installations in other contexts, so we really need to be sure it works _for everyone_.

Speaking of "working for everyone", that's the problem with the powershell approach (but with the wix one most likely too). The download of the wv2 installer needs TLS1.2 which unfortunately isn't enabled on most/many windows 7 machines _even though MS tried to enable it via an update). I think it was #2882 where we played around with that.

> UAC is not an issue anymore because ms added support for installing it with just user permissions

Btw this was a reaction to an old github comment you linked and i meant the UAC for the webview2 installer (which as your video shows doesn't exist anymore), not for the app installer...

> As a note, even down to the UAC prompt (if it remains) - that needs to be tidied up too. Show something like "${AppName} installer" or something instead of the full path to the installer.

As far as we/i know this is bound to code signing, isn't it?

**Comment by TapGhoul** [comment]
Apologies for that, I wasn't planning to make a PR until I was absolutely sure what the team wanted, and since the convo ended quickly I assumed it would just be handled by the team.

If we're okay to go ahead with ExePackage I can investigate that. I don't believe install progress can be tracked, but at the very least we can see if the bootstrapper download's progress can be tracked (the whole "This may take a while" shtick you see around) - though there's another option I'll mention responding to another section in this comment.

> * but only if it actually works fine: We had a few problems with silent installations in other contexts, so we really need to be sure it works for everyone.

Any specific known situations/test cases that I can look into for this? Discord comment links work fine, hopefully I can try to replicate them.

> Btw this was a reaction to an old github comment you linked and i meant the UAC for the webview2 installer (which as your video shows doesn't exist anymore), not for the app installer...

Oops, I thought that UAC prompt that popped up PS *was* for edge. Looking back over it, I see it's not used there. My bad, and my apologies. Though now I'm confused what was actually executing that... May need to investigate deeper, just reading the main.wxs source nothing is jumping out at me right now as to what's firing that UAC prompt off.

> Speaking of "working for everyone", that's the problem with the powershell approach (but with the wix one most likely too). The download of the wv2 installer needs TLS1.2 which unfortunately isn't enabled on most/many windows 7 machines _even though MS tried to enable it via an update). I think it was https://github.com/tauri-apps/tauri/issues/2882 where we played around with that.

Couldn't we just have Tauri download and bundle the webview2 bootstrapper directly into the installer instead? It's a fairly small exe (only 1.7MB!), so I don't see any reason not to just directly bundle it in this case if the issue is downloading - the tauri CLI already downloads wix as part of the build process anyway. Any reason not to?

> As far as we/i know this is bound to code signing, isn't it?

Oh you know what, you're probably right. I was hoping for a change to the full path, but that makes sense. I jumped the gun here, sorry 😅

**Comment by FabianLars** [comment]
> Couldn't we just have Tauri download and bundle the webview2 bootstrapper directly into the installer instead? It's a fairly small exe (only 1.7MB!),

Yeah, i was thinking about adding it in as a "windows7CompatibilityMode" or something.

> Any reason not to?

Increased installer size :P

----

PR #3076 may be interesting for the UAC stuff and machine vs user wide installs, because right around that PR was when microsoft changed the wv2 installer to not require elevated permissions.
And yes the UAC prompt back then mentioned "Edge Updater" and afaik it still does if you install webview2 with elevated permissions.

----
> Any specific known situations/test cases that I can look into for this? Discord comment links work fine, hopefully I can try to replicate them.

Not reallyyy, iirc it was mainly about silent installations not asking for elevated permissions, which shouldn't be a problem here. With microsoft themselves showing how to silently install it, i think it should be fine to just do some general tests.

**Comment by TapGhoul** [comment]
> Increased installer size :P

Not much of an increase imo and well worth the improved UX, but if you say so.

----------------------------

I'll see what I can do about this since I'm comfy enough to know what should be done, hopefully there will be an initial PR in the next couple weeks, work hours allowing.

**Comment by FabianLars** [comment]
Oh more important than the installer size (i don't care too much about it either, especially since it may be compressed even more) is that it always pulls the newest installer instead of the one you used when you created your app 2 years ago :P (btw this seems to be the default MS recommends)

But again, this is just _me_, idk what the others think, especially @lucasfernog but _i_ don't think it's worth to be the default (because of the 2 reasons above) just for windows 7 users.

**Comment by TapGhoul** [comment]
From what I understand, the evergreen bootstrapper will always pull the latest edge. I can go in and verify but that's what I got from skimming it, though I didn't dig too deep. I'll verify that before I try something like that.

**Comment by FabianLars** [comment]
> the evergreen bootstrapper will always pull the latest edge

Yes that is correct, but i mean the installer/ the 1.7mb file itself. For example the change that allowed non-admin installations was in that file (afaik).

**Comment by TapGhoul** [comment]
Ehhh... I'm on the fence as to if that's a good or bad thing, honestly.
Unless we're worried about the installer breaking altogether, generally keeping predictable behavior for a given binary I see as a positive - as long as exploits aren't a concern.
MS isn't known to abandon links until products are entirely deprecated, and even then (windows 95's windows update service stayed online far longer than it should have lmao)

**Comment by lucasfernog** [comment]
> As a note, even down to the UAC prompt (if it remains) - that needs to be tidied up too. Show something like "${AppName} installer" or something instead of the full path to the installer.

Yeah this is only fixed if you codesign the MSI.

About the issue itself, I'd love to see a different implementation. I think ExePackage isn't an option here, IIRC it changes the installer to be an exe file (because you need to use a Wix bundle instead of product) and it would be a breaking change.

And for embedding the installer, would be better if it was a CLI build option for win7 or whatever - so ideally you'd ship a separate installer for dinosaur OSes.

**Comment by TapGhoul** [comment]
It may be possible to do - the `Compressed="no"` option + `RemotePayload` child may be the key here. But I'm yet to get those working, need to find an easier way to debug the wix builds than the naiive method of just building via tauri's CLI each time. I'll see how far I can get next weekend.

See:
https://wixtoolset.org/documentation/manual/v3/xsd/wix/remotepayload.html
> Describes information about a remote file payload that is not available at the time of building the bundle. The parent must specify DownloadUrl and must not specify SourceFile when using this element.

Only thing is it requires a SHA1 hash, but `HexType` also supports empty string, so maybe? Will try next weekend.

**Comment by TapGhoul** [comment]
I think I'm gonna have to call it, there's a lot here that's over my head due to lack of familiarity with Windows.
However! I did find a useful utility for anyone else who wishes to take a crack at this: https://wixedit.github.io/

You'll wanna install the .net 3.5 SP1 runtime (even if you have 4 installed) and use the binary download, I've found their installer to generate a broken run, but it's gonna make designing the installer a lot easier. You're also gonna need WiX installed to the system, not as a set of free-floating binaries.

The problem I ran into is that it seems that Package being the outer actually prevents using ExePackage with anything, so I was unable to generate an easy test without spending far more time on this than I'm able to contribute at the moment.

See https://wixtoolset.org/documentation/manual/v3/xsd/wix/wix.html for details on that, and follow the chain up from the aforementioned RemotePayload to see how that could be structured.

In my case, I'll probably just do a thin native dialog that shows "Hey we need to install edge webview, that okay?" at startup in main. The installer seems to be a bit too much complexity for me to wrangle for the moment.
